Firestore/core/src/firebase/firestore/util/filesystem.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_FILESYSTEM_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_FILESYSTEM_H_
+
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+// High-level routines for the manipulating the filesystem. All filesystems
+// are required to implement these routines.
+
+/**
+ * Returns true if the given path exists. If the path exists and is_directory is
+ * non-null, stores whether or not the path is a directory. If the file does
+ * not exist, is_directory is unmodified.
+ */
+bool Exists(const Path& path, bool* is_directory = nullptr);
```

Comment:
I'd separate 'exists' from 'isdir'. eg:
http://google3/file/base/filesystem.h?l=263
vs
http://google3/file/base/filesystem.h?l=282

boost::filesystem also provides exists(), is_directory() as separate things.

(OTOH, stat(2) does both in one shot.)

Reply:
I'd also avoid output parameters if possible. C++17 `<filesystem>` is based on Boost and, like Rich said, has two separate functions for this. (one random idea that comes to mind is returning an enum defined like { nonexistent, file, dir }, which would convert to `bool` in a sensible manner, but I guess that's kinda hacky)

Reply:
> All three implementations are capable of determining if a file exists and if it's a directory in a single call; it seems a shame to intentionally split this despite that.

`<filesystem>` also has a bunch of additional checks like `is_regular_file`, which check both existence and specific type in one go, and [file_type enum](https://en.cppreference.com/w/cpp/filesystem/file_type) that contains `not_found`/`directory`/etc, so I guess there's always a library call that allows doing the desired check in one go. This is way overkill for our purposes, especially given the rationale you describe. Sorry, I didn't realize that the expected usage of this code is pretty specific. I'm totally fine if you revert back to `Exists` (though adding error codes seems nice).